pkg/router/controller/host_admitter.go

```diff
@@ -0,0 +1,204 @@
+func (p *HostAdmitter) HandleRoute(eventType watch.EventType, route *routeapi.Route) error {
+	if err := p.admitter(route); err != nil {
+		glog.Errorf("Route %s not admitted: %s", routeNameKey(route), err.Error())
+		p.recorder.RecordRouteRejection(route, "RouteNotAdmitted", err.Error())
+		return err
+	}
+
+	if p.restrictOwnership && len(route.Spec.Host) > 0 {
+		switch eventType {
+		case watch.Added, watch.Modified:
+			if err := p.addRoute(route); err != nil {
+				glog.Errorf("Route %s not admitted: %s", routeNameKey(route), err.Error())
+				p.recorder.RecordRouteRejection(route, "SubdomainAlreadyClaimed", err.Error())
+				return err
+			}
+
+		case watch.Deleted:
+			if subdomain := getSubdomain(route.Spec.Host); len(subdomain) > 0 {
+				p.subdomainToRoute.RemoveRoute(subdomain, route)
+			}
+		}
+	}
+
+	return p.plugin.HandleRoute(eventType, route)
+}
+
+// HandleAllowedNamespaces limits the scope of valid routes to only those that match
+// the provided namespace list.
+func (p *HostAdmitter) HandleNamespaces(namespaces sets.String) error {
+	return p.plugin.HandleNamespaces(namespaces)
+}
+
+func (p *HostAdmitter) SetLastSyncProcessed(processed bool) error {
+	return p.plugin.SetLastSyncProcessed(processed)
+}
+
+// addRoute admits routes based on subdomain ownership - returns errors if the route is not admitted.
+func (p *HostAdmitter) addRoute(route *routeapi.Route) error {
+	subdomain := getSubdomain(route.Spec.Host)
+	if len(subdomain) == 0 {
+		return nil
+	}
+
+	routeList, ok := p.subdomainToRoute[subdomain]
+	if !ok {
+		p.subdomainToRoute.InsertRoute(subdomain, route)
+		return nil
+	}
+
+	oldest := routeList[0]
+	if oldest.Namespace == route.Namespace {
+		p.subdomainToRoute.InsertRoute(subdomain, route)
+		return nil
+	}
+
+	// Route is in another namespace, land grab check here.
+	if routeapi.RouteLessThan(oldest, route) {
+		glog.V(4).Infof("Route %s cannot take subdomain %s from %s", routeNameKey(route), subdomain, routeNameKey(oldest))
+		err := fmt.Errorf("a route in another namespace holds subdomain %s and is older than %s", subdomain, route.Name)
+		p.recorder.RecordRouteRejection(route, "SubdomainAlreadyClaimed", err.Error())
+		return err
+	}
+
+	// Namespace of this route is now the proud owner of the subdomain.
+	glog.V(4).Infof("Route %s is reclaiming subdomain %s from namespace %s", routeNameKey(route), subdomain, oldest.Namespace)
+
+	// Delete all the routes belonging to the previous "owner" (namespace).
+	for idx := range routeList {
+		msg := fmt.Sprintf("a route in another namespace %s owns subdomain %s", route.Namespace, subdomain)
+		glog.V(4).Infof("Route %s not admitted: %s", routeNameKey(routeList[idx]), msg)
+		p.recorder.RecordRouteRejection(routeList[idx], "SubdomainAlreadyClaimed", msg)
+		p.plugin.HandleRoute(watch.Deleted, routeList[idx])
+	}
+
+	// And claim the subdomain.
+	p.subdomainToRoute[subdomain] = []*routeapi.Route{route}
+	return nil
+}
+
+func getSubdomain(host string) string {
```

Comment:
Consider renaming to 'getDomain', since calling it on 'foo.example.com' returns 'example.com'.